backend/src/postsDataSource.ts

```diff
@@ -0,0 +1,50 @@
+import { DataSource } from 'apollo-datasource'
+import { v4 as uuidv4 } from 'uuid'
+
+export class PostsDataSource extends DataSource {
+  posts
+  users
+  constructor (posts, users) {
+    super()
+
+    this.posts = posts
+    this.users = users
+  }
+
+  getUsers () {
+    return this.users
+  }
+
+  getPosts () {
+    return this.posts
+  }
+
+  upvote (postId, voter) {
+    const foundItem = this.getPosts().find(item => item.id === postId)
+    const foundUser = this.getUsers().find(user => user.name === voter.name)
+    if (foundItem && foundUser) {
+      if (!foundItem.voters.includes(foundUser.name)) {
+        foundItem.votes += 1
```

Comment:
You have redundant state here: `votes` and `voters`. Since `votes` is directly depending on `voters` the best idea is to implement a type resolver:
```js
export default {
  Query: { /* ... */ },
  Mutation: { /* ... */ },
  Post: {
    votes: (post) => post.voters.length
  }
```
The redundancy here could even lead to invalid state (e.g. you forget to keep `votes` up-to-date). ~shared mutable~ state is the root of all evil!